xarray/backends/api.py

```diff
@@ -352,11 +344,25 @@ def maybe_decode_store(store, lock=False):
         with close_on_error(store):
             ds = maybe_decode_store(store)
     else:
-        if engine is not None and engine != 'scipy':
-            raise ValueError('can only read file-like objects with '
-                             "default engine or engine='scipy'")
-        # assume filename_or_obj is a file-like object
-        store = backends.ScipyDataStore(filename_or_obj)
+        if engine not in [None, 'scipy', 'h5netcdf']:
+            raise ValueError('can only read bytes or file-like objects with '
+                             "engine = None, 'scipy', or 'h5netcdf'")
+        else:
+            if isinstance(filename_or_obj, bytes):
+                filename_or_obj = BytesIO(filename_or_obj)
+            # read first bytes of file-like object to determine engine
+            magic_number = filename_or_obj.read(8)
+            filename_or_obj.seek(0)
+            if magic_number.startswith(b'CDF'):
+                store = backends.ScipyDataStore(filename_or_obj,
+                                                **backend_kwargs)
+            elif magic_number.startswith(b'\211HDF\r\n\032\n'):
+                store = backends.H5NetCDFStore(filename_or_obj, group=group,
+                                               lock=lock, **backend_kwargs)
+            else:
+                print(magic_number)
+                raise ValueError("file-like object is not a netCDF file: {}"
+                                 .format(filename_or_obj))
```

Comment:
I'm a little concerned about giving users an error message about file-like objects if they passed in a `bytes` object, e.g., consider `xarray.open_dataset(b'garbage')`.

Ideally this should give a useful error message, something like: `ValueError:  b'garbage' is not a valid netCDF file, did you mean to pass a string for a path instead?`, not `ValueError: file-like object is not a netCDF file: <_io.BytesIO at 0x105663888>`.